Add temperature-control cluster xml and generate zap

jadhavrohit924 opened the pull request:
- Changes:

> Add temperature-control cluster xml and generate zap.

- Testing

> Build all-clusters-app, chip-tool

andy31415 commented:
@jadhavrohit924 lots of CI failures ... I think some file changes may be missing here.

dhrishi commented:
> @jadhavrohit924 lots of CI failures ... I think some file changes may be missing here.

Yes, right. We will fix those

jadhavrohit924 commented:
@bzbarsky-apple Please review.

jadhavrohit924 commented:
@bzbarsky-apple I have implemented server code for temperature control cluster and added PR #26566. When I read attributes from chip-tool, chip-tool received correct command data. But it got message of `Don't know how to log atribute value`. Am I missing something here?

[chip-tool.txt](https://github.com/project-chip/connectedhomeip/files/11476310/chip-tool.txt)

bzbarsky-apple commented:
@jadhavrohit924 How does this PR interact with the XML being added in https://github.com/project-chip/connectedhomeip/pull/26560 ?

Looking into your chip-tool issue

bzbarsky-apple commented:
> Am I missing something here?

It's a chip-tool bug.  See https://github.com/project-chip/connectedhomeip/pull/26577

jadhavrohit924 commented:
@bzbarsky-apple 
> It's a chip-tool bug. See 26577.

Actually, I tried enabling client cluster in controller-clusters.zap. But it didn’t  work.

bzbarsky-apple commented:
> Actually, I tried enabling client cluster in controller-clusters.zap. But it didn’t work.

I don't know exactly what you did, but if I enable the client cluster and then rerun codegen, DataModelLogger picks up these new attributes.  That's without https://github.com/project-chip/connectedhomeip/pull/26577.  With that it does not matter and it all just works.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:
```diff
@@ -0,0 +1,52 @@
+    <define>TEMPERATURE_CONTROL_CLUSTER</define>
+    <client tick="false" init="false">true</client>
+    <server tick="false" init="false">true</server>
+    <attribute side="server" code="0x0000" define="TEMP_SETPOINT" type="INT16S" min="0x8000" max="0x7fff" writable="false"  optional="true">TemperatureSetpoint</attribute>
```
Please take out the min and max bits for all of these readonly attributes: not only do they not match the spec, we're not using the values for anything.  And having them as positive hex values for negative things is fairly confusing.

jadhavrohit924 replied:
Removed

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:
```diff
@@ -0,0 +1,52 @@
+
+    <command source="client" code="00" name="SetTemperature" optional="false">
+        <description>Set Temperature</description>
+        <arg name="TargetTemperature" type="INT16S" min="MIN_TEMP" max="MAX_TEMP"/>
+        <arg name="TargetTemperatureLevel" type="INT8U" />
```
Both of those args are optional, no?

jadhavrohit924 replied:
Yes updated

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:
```diff
@@ -0,0 +1,52 @@
+    <attribute side="server" code="0x0004" define="CURRENT_TEMP_LEVEL_IND" type="INT8U" min="0x00" max="0x1f" writable="false" optional="true">CurrentTemperatureLevelIndex</attribute>
+    <attribute side="server" code="0x0005" define="SUPPORTED_TEMP_LEVELS" type="ARRAY" entryType="TemperatureLevelStruct" writable="false" optional="true">SupportedTemperatureLevels</attribute>
+
+    <command source="client" code="00" name="SetTemperature" optional="false">
```
```suggestion
    <command source="client" code="0x0" name="SetTemperature" optional="false">
```

jadhavrohit924 commented on src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:
```diff
@@ -0,0 +1,52 @@
+    <command source="client" code="0x00" name="SetTemperature" optional="false">
+        <description>Set Temperature</description>
+        <arg name="TargetTemperature" type="INT16S" min="MIN_TEMP" max="MAX_TEMP" optional="true"/>
+        <arg name="TargetTemperatureLevel" type="INT8U" optional="true"/>
```
@bzbarsky-apple These shouldn’t be set to optional. Both of there args are feature dependent and cluster must have any one feature enabled. So if we set these args to optional, chip-tool will generate command like this `chip-tool temperaturecontrol set-temperature destination-id endpoint-id-ignored-for-group-commands`

jadhavrohit924 replied:
We can definitely pass `TargetTemperature` and `TargetTemperatureLevel` to the command. But shouldn’t we have atleast one argument mandatory ( whichever feature we are enabling by default )?

bzbarsky-apple replied:
They are feature dependent, which means that any given cluster will have one or the other of them missing.  Which means the command struct must be able to represent "missing" so it does not send garbage on the wire.  Which is why they have to be optional.

The XML does not have a way to indicate "at least one of these is mandatory".  You have to define them both as optional and then consumers have to make sure they pass one or the other as needed.

jadhavrohit924 replied:
Okay. Thanks.

bzbarsky-apple commented:
@jadhavrohit924 Merge conflicts; needs a regen.

woody-apple commented:
@jadhavrohit924 Would you mind taking a peek at the merge conflicts?

bzbarsky-apple commented:
@woody-apple There are no merge conflicts; this is just blocked on non-Apple review.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:
Note that this is likely to need significant changes based on the outcome of https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/7040.  Please make sure those changes are either made before this lands or tracked in a cert blocker issue after this lands.

dhrishi replied:
@bzbarsky-apple Anything works. I can also also open a SDK issue and assign to @jadhavrohit924 to track this

dhrishi replied:
https://github.com/project-chip/connectedhomeip/issues/26646

marktrayer commented on src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:
```diff
@@ -0,0 +1,52 @@
+
+  <cluster>
+    <name>Temperature Control</name>
+    <domain>Home &amp; Appliances</domain>
```
Please set domain to just: Appliances

jadhavrohit924 replied:
Done

bzbarsky-apple commented:
This may need a rebase to tip to fix the ZAP CI issue, since it got updated while tip was broken... @jadhavrohit924

dhrishi commented:
@jadhavrohit924  `Smoke test - Android / Smoke Run - Android` in CI is failing :
```
INFO    2023-05-19 02:12:02 INFO    Writing new data to: gen/src/controller/data_model/jni/FaultInjectionClient-InvokeSubscribeImpl.cpp
INFO    2023-05-19 02:12:02 CRITICAL expected and generated files do not match.
INFO    2023-05-19 02:12:02 CRITICAL    'jni/TemperatureControlClient-InvokeSubscribeImpl.cpp' was generated but not expected
INFO    2023-05-19 02:12:02 CRITICAL    'jni/TemperatureControlClient-ReadImpl.cpp' was generated but not expected
INFO    [60/1418] ACTION //src/controller/data_model:data_model_zapgen_zap_pregen(//build/toolchain/android:android_arm64)
INFO    [61/1418] ACTION //src/controller/java:tlv__kotlinc(//build/toolchain/android:android_arm64)
INFO    warning: flag is not supported by this version of the compiler: -Xlint:all
INFO    warning: flag is not supported by this version of the compiler: -Xlint:deprecation
INFO    ninja: build stopped: subcommand failed.
```
cc @bzbarsky-apple

bzbarsky-apple commented:
@dhrishi @jadhavrohit924 Those two files need to be added to the outputs list in `src/controller/data_model/BUILD.gn`.  Search for `jni/` in there and note the list is in alphabetical order.

jadhavrohit924 commented:
@bzbarsky-apple Thanks for help :)

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:
```diff
@@ -0,0 +1,52 @@
+  <struct name="TemperatureLevelStruct">
+    <cluster code="0x0056"/>
+    <item fieldId="0" name="Label" type="char_string" length="64" isNullable="false" writable="false"/>
+    <item fieldId="1" name="TempLevel" type="int8u" isNullable="false" writable="false"/>
```
I missed this, but this does not match the spec....  I guess I'll fix it up in a separate PR, but @jadhavrohit924 please be more careful in the future.

bzbarsky-apple replied:
https://github.com/project-chip/connectedhomeip/pull/26691

dhrishi replied:
@jadhavrohit924 Please take a note
